keps/sig-node/4816-dra-prioritized-list/README.md

```diff
@@ -0,0 +1,1048 @@
+
+```yaml
+apiVersion: resource.k8s.io/v1alpha4
+kind: ResourceClaim
+metadata:
+  name: device-consumer-claim
+spec:
+  devices:
+    requests:
+    - name: nic
+      deviceClassName: rdma-nic
+    - name: gpu
+      firstOf:
+      - name: big-gpu
+        deviceClassName: big-gpu
+      - name: mid-gpu
+        deviceClassName: mid-gpu
+      - name: small-gpu
+        deviceClassName: small-gpu
+        count: 2
+    constraints:
+    - requests: ["nic", gpu"]
+      matchAttribute:
+      - dra.k8s.io/pcieRoot
+    config:
+    - requests: ["small-gpu"]
+      opaque:
+        driver: gpu.acme.example.com
+        parameters:
+          apiVersion: gpu.acme.example.com/v1
+          kind: GPUConfig
+          mode: multipleGPUs
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  name: device-consumer
+spec:
+  resourceClaims:
+  - name: "gpu-and-nic"
+    resourceClaimName: device-consumer-claim
+  containers:
+  - name: workload
+    image: my-app
+    command: ["/bin/program"]
+    resources:
+      requests:
+        memory: "64Mi"
+        cpu: "250m"
+      limits:
+        memory: "128Mi"
+        cpu: "500m"
+      claims:
+      - name: "gpu-and-nic"
+        request: "gpu" # the 'nic' request is pod-level, no need to attach to container
+```
+
+There are a few things to note here. First, the "nic" request is listed with a
+`deviceClassName`, because it has no alternative request types. The "gpu"
+request could be met by several different types of GPU, in the listed order of
+preference. Each of those is a separate `DeviceRequest`, with both a
+`deviceClassName` and also its own name. The fact that these subrequests also
+have their own names allows us to apply constraints or configuration to
+specific, individual subrequests, in the event that it is the chosen
+alternative. In this example, the "small-gpu" choice requires a configuration
+option that the other two choices do not need. Thus, if the resolution of the
+"gpu" request is made using the "small-gpu" subrequest, then that configuration
+will be attached to the allocation. Otherwise, it will not.
+
+Similarly, for `Constraints`, the list of requests can include the main request
+name ("gpu" in this case), in which case the constraint applies regardless of
+which alternative is chosen. Or, it can include the subrequest name, in which
+case that constraint only applies if that particular subrequest is chosen.
+
+In the PodSpec, however, the subrequest names are not valid. Only the main
+request name may be used.
+
+### Resource Quota
+
+ResourceQuota will be enforced such that the user must have quota for each
+`DeviceRequest` under every `FirstOf`. Thus, this "pick one" behavior cannot be
+used to circumvent quota. This reduces the usefulness of the feature, as it
+means it will not serve as a quota management feature. However, the primary goal
+of the feature is about flexibility across clusters and obtainability of
+underlying devices, not quota management.
```

Comment:
Hopefully kueue can help with that.
cc @kannon92

Reply:
@johnbelamaric could we call out Kueue as a beta requirement? I’d really like to Kueue usecase in the requirements for DRA features.

WDYT @alculquicondor?

Reply:
But Kueue is a great way to get feedback on how useful this feature is for future users?

Reply:
I think of this as the k8s projects dogfooding their own APIs so we can see what is useful or not. But I do take your point that Kueue is not in tree.